src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs

```diff
@@ -1,60 +1,97 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-using System.Collections.Generic;
+using System;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Template;
+using Microsoft.Extensions.Options;
 
 namespace Microsoft.AspNetCore.Dispatcher
 {
     public class DispatcherMiddleware
     {
+        private readonly DispatcherOptions _options;
         private readonly RequestDelegate _next;
 
-        public DispatcherMiddleware(RequestDelegate next)
+        public DispatcherMiddleware(IOptions<DispatcherOptions> options, RequestDelegate next)
         {
+            _options = options.Value;
             _next = next;
         }
 
         public async Task Invoke(HttpContext httpContext)
         {
-            var dictionary = new Dictionary<string, DispatcherFeature>
+            foreach (var entry in _options.DispatcherEntryList)
             {
+                var parsedTemplate = entry.RouteTemplate;
+                var defaults = GetDefaults(parsedTemplate);
+                var templateMatcher = new TemplateMatcher(parsedTemplate, defaults);
+                var values = new RouteValueDictionary();
+
+                foreach (var endpoint in entry.Endpoints)
                 {
-                    "/example",
-                    new DispatcherFeature
+                    if (templateMatcher.TryMatch(httpContext.Request.Path, values))
+                    {
+                        if (!CompareRouteValues(values, endpoint.RequiredValues))
                         {
-                            Endpoint = new DispatcherEndpoint("example"),
-                            RequestDelegate = async (context) =>
-                            {
-                                await context.Response.WriteAsync("Hello from the example!");
-                            }
+                            values.Clear();
                         }
-                },
-                {
-                    "/example2",
-                    new DispatcherFeature
+
+                        else
                         {
-                            Endpoint = new DispatcherEndpoint("example2"),
-                            RequestDelegate = async (context) =>
+                            var dispatcherFeature = new DispatcherFeature
                             {
-                                await context.Response.WriteAsync("Hello from the second example!");
-                            }
+                                Endpoint = endpoint,
+                                RequestDelegate = endpoint.RequestDelegate
+                            };
+
+                            httpContext.Features.Set<IDispatcherFeature>(dispatcherFeature);
```

Comment:
You also need to a `break` here. You don't want to process multiple entries - we should move on once we have a match